Equilobe.DailyReport.SL/ReportService.cs

```diff
@@ -12,27 +13,97 @@ public class ReportService : IReportService
         public void UpdateDashboardData(long instanceId)
         {
-            UpdateAtlassianUsers(instanceId);
+            JiraRequestContext = GetJiraRequestContext(instanceId);
+
+            SyncAtlassianUsers(instanceId);
+            SyncAtlassianWorklogs(instanceId);
         }
         #endregion
 
         #region Helpers
-        private void UpdateAtlassianUsers(long instanceId)
+        private void SyncAtlassianWorklogs(long instanceId)
         {
-            var jiraRequestContext = GetJiraRequestContext(instanceId);
-            var users = JiraService.GetAllUsers(jiraRequestContext)
-                .Select(p => ToAtlassianUser(p, instanceId))
-                .ToList();
+            SyncDeletedWorklogs(instanceId);
+            SyncUpdatedWorklogs(instanceId);
+        }
+
+        private void SyncDeletedWorklogs(long instanceId)
+        {
+            var deletedWorklogsIds = JiraService.GetDeletedWorklogsIds(JiraRequestContext, DateTime.UtcNow.AddMonths(-7));
+
+            if (deletedWorklogsIds == null)
+                return;
 
             using (var db = new ReportsDb())
             {
-                var dbUsers = db.AtlassianUser
+                var dbWorklogs = db.AtlassianWorklog
                     .Where(p => p.InstalledInstanceId == instanceId)
                     .ToList();
 
+                foreach (var worklogId in deletedWorklogsIds)
+                {
+                    var dbWorklog = dbWorklogs.SingleOrDefault(p => p.JiraWorklogId == worklogId);
+
+                    if (dbWorklog != null)
+                        db.AtlassianWorklog.Remove(dbWorklog);
+                }
+
+                db.SaveChanges();
+            }
+        }
+
+        private void SyncUpdatedWorklogs(long instanceId)
+        {
+            var worklogs = GetAtlassianWorklogs(instanceId);
+
+            using (var db = new ReportsDb())
+            {
+                var dbWorklogs = db.AtlassianWorklog
+                    .Where(p => p.InstalledInstanceId == instanceId)
+                    .ToList();
+
+                foreach (var worklog in worklogs)
+                {
+                    var dbWorklog = dbWorklogs.Where(p => p.JiraWorklogId == worklog.JiraWorklogId).SingleOrDefault();
+
+                    if (dbWorklog == null)
+                        db.AtlassianWorklog.Add(worklog);
+                    else if (dbWorklog.UpdatedAt != worklog.UpdatedAt)
+                        UpdateDbWorklog(dbWorklog, worklog);
+                }
+
+                db.SaveChanges();
+            }
+        }
+
+        private List<AtlassianWorklog> GetAtlassianWorklogs(long instanceId)
+        {
+            var users = GetAllAtlassianUsers(instanceId);
+            var userKeys = users
+                .Select(p => p.Key)
+                .ToList();
+
+            var issueWorklogs = JiraService.GetWorklogsForMultipleUsers(JiraRequestContext, userKeys, DateTime.UtcNow.AddMonths(-1), DateTime.UtcNow);
```

Comment:
Let's only use the last month in case the date of the last sync is null. Otherwise, always get the worklogs that where either added after that date, either updated after that date.